libcaja-private/caja-file.c

```diff
@@ -3160,6 +3162,192 @@ compare_by_full_path (CajaFile *file_1, CajaFile *file_2)
 	return compare_by_display_name (file_1, file_2);
 }
 
+/* prev_extension_segment:
+ * @basename The basename of a file
+ * @rem_chars A pointer to the amount of remaining characters
+ *
+ * Finds the next segment delimiter to the left. A starting character of '.' is
+ * set to '\0'.
+ *
+ * Return value: The start of the previous segment (right of the dot) or
+ * basename if there are none remaining.
+ */
+static char *
+prev_extension_segment (char *basename, int *rem_chars)
+{
+	if (*basename == '.') {
+		*basename = 0;
+		basename--;
+		(*rem_chars)--;
+	}
+
+	while (*rem_chars > 0 && *basename != '.') {
+		(*rem_chars)--;
+		basename--;
+	}
+
+	return basename + 1;
+}
+
+/* is_valid_extension_segment:
+ * @segment Part of a modifiable zero-terminated string
+ * @segment_index The index of the current segment
+ *
+ * Uses a heuristic to identify valid file extensions.
+ *
+ * Return value: Whether the segment is part of the file extension.
+ */
+static gboolean
+is_valid_extension_segment (const char *segment, int segment_index)
+{
+	gboolean result;
+	gboolean has_letters;
+	char c;
+	int n;
+	switch (segment_index) {
+		case 0:
+			/* extremely long segments are probably not part of the extension */
+			result = strlen (segment) < 20;
+			break;
+		default:
+			has_letters = FALSE;
+			n = 0;
+			while (TRUE) {
+				c = *(segment + n);
+				if (c == '\0') {
+					result = has_letters;
+					break;
+				}
+				/* allow digits if there are also letters */
+				else if (isalpha (c)) {
+					has_letters = TRUE;
+				}
+				/* fail if it is neither digit nor letter */
+				else if (!isdigit (c)) {
+					result = FALSE;
+					break;
+				}
+
+				/* only accept short segments */
+				if (n > 3) {
```

Comment:
This `n` value is confusing - I'm assuming that since you're inside a `do`, the whole thing is added 1? I have some `foo.debian.tar.xz` and some `foo.orig.tar.xz` files: I can see the `orig` part in the extensions column, but not the `debian`.

I guess I don't mind where the cut-off is since if you're using `.debian.tar.xz` to do anything meaningful you should probably be in a terminal :smile_cat: - anyway, my issue is that this could use a `#define` value for the number of characters that still get treated as an extension, say for example `#define SORT_BY_EXTENSION_MAX_LENGTH`, and then here you can do
```c
if (n > SORT_BY_EXTENSION_MAX_LENGTH - 1) { ... }
```
(I also noticed that there's a similar `while (segment_index < 4);` below, so a defined macro number will make sense for future reference/debugging/updating)

Reply:
> This n value is confusing - I'm assuming that since you're inside a do, the whole thing is added 1?

You are absolutely correct. I will change it to `char_offset` to properly reflect its purpose.

> I have some foo.debian.tar.xz and some foo.orig.tar.xz files: I can see the orig part in the extensions column, but not the debian.

Neither should actually show (maximum length of following segments is 3). It is a bug, which I will fix in an instant. Now, I would be more than happy if someone actually came up with a heuristic that would work better. But I fear dots are used for other purposes too. Another approach would be to use a list of known extension segments (`tar`, `pkg`, `debian`). But it could be very time-consuming.

I will replace the hard-coded constants with macros.